app/src/main/java/org/wikipedia/search/SearchResultsViewModel.kt

```diff
@@ -0,0 +1,171 @@
+package org.wikipedia.search
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import androidx.paging.*
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.withContext
+import org.wikipedia.WikipediaApp
+import org.wikipedia.database.AppDatabase
+import org.wikipedia.dataclient.ServiceFactory
+import org.wikipedia.dataclient.WikiSite
+import org.wikipedia.dataclient.mwapi.MwQueryResponse
+import org.wikipedia.util.StringUtil
+import java.util.*
+
+class SearchResultsViewModel : ViewModel() {
+
+    private val batchSize = 20
+    private val delayMillis = 200L
+    private val totalResults = mutableListOf<SearchResult>()
+    var resultsCount = mutableListOf<Int>()
+    var searchTerm: String? = null
+    var languageCode: String? = null
+
+    @OptIn(FlowPreview::class) // TODO: revisit if the debounce method changed.
+    val searchResultsFlow = Pager(PagingConfig(pageSize = batchSize, initialLoadSize = batchSize)) {
+        SearchResultsPagingSource(searchTerm, languageCode, resultsCount, totalResults)
+    }.flow.debounce(delayMillis).map { pagingData ->
+        pagingData.filter { searchResult ->
+            totalResults.find { it.pageTitle.prefixedText == searchResult.pageTitle.prefixedText } == null
+        }.map {
+            totalResults.add(it)
+            it
+        }
+    }.cachedIn(viewModelScope)
+
+    fun clearResults() {
+        resultsCount.clear()
+        totalResults.clear()
+    }
+
+    class SearchResultsPagingSource(
+            val searchTerm: String?,
+            val languageCode: String?,
+            var resultsCount: MutableList<Int>?,
+            var totalResults: MutableList<SearchResult>?
+    ) : PagingSource<MwQueryResponse.Continuation, SearchResult>() {
+
+        private val batchSize = 20
+        private var prefixSearch = true
+
+        override suspend fun load(params: LoadParams<MwQueryResponse.Continuation>): LoadResult<MwQueryResponse.Continuation, SearchResult> {
+            return try {
+                // The default offset is 0 but we send the initial offset from 1 to prevent showing the same talk page from the results.
+                if (searchTerm.isNullOrEmpty() || languageCode.isNullOrEmpty()) {
+                    return LoadResult.Page(emptyList(), null, null)
+                }
+
+                val wikiSite = WikiSite.forLanguageCode(languageCode)
+                var response: MwQueryResponse? = null
+                var readingListSearch = SearchResults()
+                var historySearch = SearchResults()
+                if (prefixSearch) {
+                    if (searchTerm.length > 2) {
+                        readingListSearch = withContext(Dispatchers.IO) {
```

Comment:
It looks like we're doing `readingListSearch` and `historySearch` sequentially, not concurrently. Can we perform them both (along with prefixSearch) concurrently?

Reply:
Done!